src/tensors/tensor_operators.h

```diff
@@ -24,15 +24,20 @@
 
 namespace marian {
 
-template <typename InIt, typename OutIt>
-void copy(Ptr<Backend> backend, const InIt beg, const InIt end, OutIt it) {
 #ifdef CUDA_FOUND
+template <typename InIt, typename OutIt>
+void copy(Ptr<Backend>& backend, const InIt beg, const InIt end, OutIt it) {
   if(backend->getDeviceId().type == DeviceType::gpu)
     gpu::copy(backend, beg, end, it);
   else
-#endif
     std::copy(beg, end, it);
 }
+#else
+template <typename InIt, typename OutIt>
+void copy(Ptr<Backend>& , const InIt beg, const InIt end, OutIt it) {
```

Comment:
Couldn't you just say

backend;

in an #else branch?

Reply:
I don't like no-op code just to deal with compiler warnings, especially since I'm not sure if that's actually a no-op, or whether the compiled code will still access memory (and potentially screw up caching). However, I agree that code duplication is undesirable. The proper way (on those compilers that support it, is through `__attribute__(unused)` (or `[[maybe_unused]]`). See macro MAYBE_UNUSED in src/common/definitions.h.

Reply:
We do do this at other places though, so it wouldn't be unusual in this codebase. MAYBE_UNUSED is good.